experimental/formatted_string/src/formatted_string.rs

```diff
@@ -0,0 +1,251 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::FormattedStringBuilderError;
+use alloc::borrow::ToOwned;
+use alloc::boxed::Box;
+use alloc::string::String;
+use alloc::vec::Vec;
+use core::str;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum LocationInPart {
+    Begin,
+    Extend,
+}
+
+pub trait FormattedString<F: Copy, const L: usize>: AsRef<str> {
+    fn fields_at(&self, pos: usize) -> [F; L] {
+        self.annotation_at(pos).map(|(_, field)| field)
+    }
+
+    fn is_field_start(&self, pos: usize, level: usize) -> bool {
+        assert!(level < L);
+        let (location, _) = self.annotation_at(pos)[level];
+        location == LocationInPart::Begin
+    }
+
+    #[doc(hidden)]
+    // This is used to make the builder more efficient; clients should
+    // use fields_at and is_field_start.
+    fn annotation_at(&self, pos: usize) -> &[(LocationInPart, F); L];
+}
+
+impl<F: Copy> FormattedString<F, 0> for &str {
+    fn annotation_at(&self, _pos: usize) -> &[(LocationInPart, F); 0] {
+        // Yay we can return dangling references for singleton types!
+        &[]
+    }
+}
+
+/// A string with L levels of formatting annotations. The name refers
+/// to the fact that the trait FormattedString is also implemented for
+/// &str, but with 0 levels of annotations there isn't much formatting.
+#[derive(Debug, PartialEq)]
+pub struct ActualFormattedString<F: Copy, const L: usize> {
```

Comment:
Suggestion: I think we should have a single FormattedString[Builder] type, rather than having this intermediate one.

Reply:
Potential type names:

- Delete `ActualFormattedString`
- `FormattedStringBuilder` --> `FormattedString`
- `FormattedString` --> `FormattedStringLike`